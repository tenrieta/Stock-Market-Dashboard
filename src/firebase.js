import { initializeApp } from "firebase/app";
import { getAuth } from "firebase/auth";
import { getFirestore, doc, setDoc, getDoc } from "firebase/firestore";

const firebaseConfig = {
  apiKey: process.env.REACT_APP_FIREBASE_API_KEY,
  authDomain: process.env.REACT_APP_FIREBASE_AUTH_DOMAIN,
  projectId: process.env.REACT_APP_FIREBASE_PROJECT_ID,
  storageBucket: process.env.REACT_APP_FIREBASE_STORAGE_BUCKET,
  messagingSenderId: process.env.REACT_APP_FIREBASE_MESSAGING_SENDER_ID,
  appId: process.env.REACT_APP_FIREBASE_APP_ID,
};

// Initialize Firebase
const firebaseApp = initializeApp(firebaseConfig);

export const auth = getAuth();
export const db = getFirestore();

export const saveFirestore = async (newData) => {
  try {
    await setDoc(doc(db, "users", auth.currentUser.uid), newData);
  } catch (e) {
    console.error("Error adding document: ", e);
  }
};

export const readFirestore = async () => {
  try {
    const docRef = doc(db, "users", auth.currentUser.uid);
    const docSnap = await getDoc(docRef);
    if (docSnap.exists()) {
      return docSnap.data();
    }
    return { error: "User not found..." };
  } catch (error) {
    return { error };
  }
};

export default firebaseApp;
