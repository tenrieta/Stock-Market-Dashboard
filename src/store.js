/* eslint-disable no-underscore-dangle */
import { applyMiddleware, compose, createStore } from "redux";
import thunk from "redux-thunk";
import rootReducer from "./reducers";

const store = createStore(
  rootReducer,
  compose(
    applyMiddleware(thunk), // Thunk is necessary for async action creators
    window.__REDUX_DEVTOOLS_EXTENSION__
      ? window.__REDUX_DEVTOOLS_EXTENSION__()
      : (f) => f // This is to make the Redux for Chrome extension work in an incognito window with disabled Redux extension
  )
);

export default store;
