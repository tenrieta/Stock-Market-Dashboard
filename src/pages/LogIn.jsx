import React from "react";

// Bootstrap
import Container from "react-bootstrap/Container";
import Row from "react-bootstrap/Row";
import Col from "react-bootstrap/Col";
import Card from "react-bootstrap/Card";
import Form from "react-bootstrap/Form";
import Button from "react-bootstrap/Button";
import Alert from "react-bootstrap/Alert";

// Firebase
import { signInWithEmailAndPassword } from "firebase/auth";
import { auth } from "../firebase";

const LogIn = function LogIn() {
  // Local state for Form Data
  const [formData, setFormData] = React.useState({
    formError: null,
    formValidated: false,
    email: "",
    password: "",
  });

  // Change handler which updates local state
  // In order to make it universal for all fields, we dynamically use [fieldName]: fieldValue
  const handleChange = (event) => {
    const fieldName = event.target.name;
    const fieldValue = event.target.value;
    setFormData((prevState) => ({ ...prevState, [fieldName]: fieldValue }));
  };

  // Submit handler
  const handleSubmit = (event) => {
    // First preventing default form behavior
    event.preventDefault();

    // Validation with Bootstrap
    setFormData((prevState) => ({ ...prevState, formValidated: true }));
    if (event.currentTarget.checkValidity() === false) {
      event.stopPropagation();
      return null;
    }

    // Sending data to Firebase
    // When Firebase approves and logs the user in, user data is dispatched to Redux from App.jsx, where we have a listener onAuthStateChanged
    signInWithEmailAndPassword(auth, formData.email, formData.password)
      // If there is an error, we show it in the form
      .catch((firebaseError) => {
        setFormData((prevState) => ({
          ...prevState,
          formValidated: false,
          formError: `There was an error: ${firebaseError.code}.`,
        }));
        // We also set a timer to clean the error message and form validation
        setTimeout(() => {
          setFormData({
            ...formData,
            formError: "",
          });
        }, 3000);
      });

    return null;
  };

  return (
    <Container>
      <Row className="justify-content-md-center">
        <Col className="col-lg-6 col-md-10 mx-auto">
          <Card className="mb-4">
            <Card.Body>
              <h2 className="text-center mb-4">Log In</h2>
              {formData.formError && (
                <Alert variant="danger">{formData.formError}</Alert>
              )}
              <Form
                noValidate
                validated={formData.formValidated}
                onSubmit={handleSubmit}
              >
                <Form.Group className="mb-3">
                  <Form.Label htmlFor="email">Email</Form.Label>
                  <Form.Control
                    type="email"
                    required
                    value={formData.email}
                    onChange={handleChange}
                    name="email"
                    id="email"
                  />
                </Form.Group>
                <Form.Group className="mb-3">
                  <Form.Label htmlFor="password">Password</Form.Label>
                  <Form.Control
                    type="password"
                    value={formData.password}
                    onChange={handleChange}
                    name="password"
                    id="password"
                  />
                </Form.Group>
                <div className="text-center">
                  <Button
                    disabled={formData.email === "" || formData.password === ""}
                    variant="primary"
                    id="submit"
                    type="submit"
                    name="submit"
                  >
                    Log In
                  </Button>
                </div>
              </Form>
            </Card.Body>
          </Card>
        </Col>
      </Row>
    </Container>
  );
};
export default LogIn;
