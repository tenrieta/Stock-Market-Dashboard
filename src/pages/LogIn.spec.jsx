import React from "react";
import "@testing-library/jest-dom";
import renderer from "react-test-renderer";
import { render, fireEvent } from "@testing-library/react";
import { Provider } from "react-redux";
import { signInWithEmailAndPassword, signOut } from "firebase/auth";
import { auth } from "../firebase";
import store from "../store";
import LogIn from "./LogIn";

describe("SNAPSHOT TESTS", () => {
  test("SignUp snapshot renders", () => {
    const component = renderer.create(
      <Provider store={store}>
        <LogIn />
      </Provider>
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe("UNIT TESTS", () => {
  test("Log In form is usable.", () => {
    const { getByRole, getByLabelText } = render(
      <Provider store={store}>
        <LogIn />
      </Provider>
    );

    const emailInput = getByLabelText(/Email/i);
    expect(emailInput).toBeInTheDocument();
    expect(emailInput).toHaveAttribute("type", "email");

    const passwordInput = getByLabelText(/Password/i);
    expect(passwordInput).toBeInTheDocument();
    expect(passwordInput).toHaveAttribute("type", "password");

    const submitButton = getByRole("button", { type: "submit" });
    expect(submitButton).toBeInTheDocument();
    expect(submitButton).toHaveAttribute("type", "submit");
  });

  test("Log In button is disabled if form is loading Firebase status, or if email or password are not filled out.", () => {
    const { getByLabelText, getByRole } = render(
      <Provider store={store}>
        <LogIn />
      </Provider>
    );
    const emailInput = getByLabelText(/Email/i);
    const passwordInput = getByLabelText(/Password/i);
    const submitButton = getByRole("button", { type: "submit" });
    expect(submitButton).toHaveAttribute("disabled");
    fireEvent.change(emailInput, { target: { value: "jest@example.com" } });
    fireEvent.change(passwordInput, { target: { value: "abc123" } });
    expect(submitButton).not.toHaveAttribute("disabled");
  });
});

describe("INTEGRATION TESTS", () => {
  test("Sign in should throw error on wrong credentials.", async () => {
    let error = "";
    try {
      await signInWithEmailAndPassword(auth, "random@email.com", "1");
    } catch (err) {
      error = err.toString();
    }
    expect(error).toMatch(/user-not-found/);
  });

  test("Sign in should return logged in user from Firebase on correct credentials. ", async () => {
    try {
      await signInWithEmailAndPassword(auth, "jest@example.com", "abc123");
    } catch (err) {
      console.log(err.toString());
    }
    expect(auth.currentUser).toBeTruthy();
    expect(auth.currentUser.email).toBe("jest@example.com");
  });

  test("Sign out should work with Firebase.", async () => {
    await signInWithEmailAndPassword(auth, "jest@example.com", "abc123");
    expect(auth.currentUser).toBeTruthy();
    await signOut(auth);
    expect(auth.currentUser).toBeFalsy();
  });
});
