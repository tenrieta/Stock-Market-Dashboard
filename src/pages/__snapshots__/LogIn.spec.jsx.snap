// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SNAPSHOT TESTS SignUp snapshot renders 1`] = `
<div
  className="container"
>
  <div
    className="justify-content-md-center row"
  >
    <div
      className="col-lg-6 col-md-10 mx-auto col"
    >
      <div
        className="mb-4 card"
      >
        <div
          className="card-body"
        >
          <h2
            className="text-center mb-4"
          >
            Log In
          </h2>
          <form
            className=""
            noValidate={true}
            onSubmit={[Function]}
          >
            <div
              className="mb-3"
            >
              <label
                className="form-label"
                htmlFor="email"
              >
                Email
              </label>
              <input
                className="form-control"
                id="email"
                name="email"
                onChange={[Function]}
                required={true}
                type="email"
                value=""
              />
            </div>
            <div
              className="mb-3"
            >
              <label
                className="form-label"
                htmlFor="password"
              >
                Password
              </label>
              <input
                className="form-control"
                id="password"
                name="password"
                onChange={[Function]}
                type="password"
                value=""
              />
            </div>
            <div
              className="text-center"
            >
              <button
                className="btn btn-primary"
                disabled={true}
                id="submit"
                name="submit"
                type="submit"
              >
                Log In
              </button>
            </div>
          </form>
        </div>
      </div>
    </div>
  </div>
</div>
`;
